File: MyFarmProduce.Domain/Entities/Product.cs
using MyFarmProduce.Common.Enums;

namespace MyFarmProduce.Domain.Entities;

public class Product
{
    public int Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;

    public int CategoryId { get; set; }
    public Category? Category { get; set; }

    public ProductUnit Unit { get; set; }
    public decimal Price { get; set; }

    // StockQty is intentionally read-only from the outside. The only ways to
    // change it are ReduceStock / Restock, which keeps the value consistent
    // and lets EF Core change-tracking observe every mutation correctly.
    public int StockQty { get; private set; }

    public bool IsAvailable { get; set; }
    public string? ImageUrl { get; set; }

    public ICollection<OrderItem> OrderItems { get; set; } = new List<OrderItem>();

    /// <summary>
    /// Decrease available stock by <paramref name="quantity"/>.
    /// Throws if the quantity is not positive or exceeds current stock.
    /// </summary>
    public void ReduceStock(int quantity)
    {
        if (quantity <= 0)
            throw new ArgumentOutOfRangeException(nameof(quantity), "Quantity must be greater than zero.");

        if (quantity > StockQty)
            throw new InvalidOperationException(
                $"Insufficient stock for '{Name}'. Available: {StockQty}, requested: {quantity}.");

        StockQty -= quantity;
    }

    /// <summary>
    /// Increase available stock by <paramref name="quantity"/>.
    /// Throws if the quantity is not positive.
    /// </summary>
    public void Restock(int quantity)
    {
        if (quantity <= 0)
            throw new ArgumentOutOfRangeException(nameof(quantity), "Quantity must be greater than zero.");

        StockQty += quantity;
    }

    /// <summary>
    /// Sets the initial stock level. Intended for entity creation / seeding only.
    /// </summary>
    public void SetInitialStock(int quantity)
    {
        if (quantity < 0)
            throw new ArgumentOutOfRangeException(nameof(quantity), "Initial stock cannot be negative.");

        StockQty = quantity;
    }
}

File: domain/src/main/java/com/myfarmproduce/domain/entity/Product.java
package com.myfarmproduce.domain.entity;

import com.myfarmproduce.domain.enums.ProductUnit;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "products")
@Getter
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Setter
    @Column(nullable = false, length = 150)
    private String name = "";

    @Setter
    @Column(length = 1000)
    private String description = "";

    @Setter
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id", nullable = false)
    private Category category;

    @Setter
    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ProductUnit unit;

    @Setter
    @Column(nullable = false, precision = 18, scale = 2)
    private BigDecimal price;

    // Intentionally no public setter: the only ways to change stock are
    // reduceStock/restock/setInitialStock, which keep the value consistent.
    @Column(name = "stock_qty", nullable = false)
    private int stockQty;

    @Setter
    @Column(name = "is_available", nullable = false)
    private boolean available;

    @Setter
    @Column(name = "image_url", length = 500)
    private String imageUrl;

    @OneToMany(mappedBy = "product")
    private List<OrderItem> orderItems = new ArrayList<>();

    /**
     * Decrease available stock by {@code quantity}.
     * Throws if the quantity is not positive or exceeds current stock.
     */
    public void reduceStock(int quantity) {
        if (quantity <= 0)
            throw new IllegalArgumentException("Quantity must be greater than zero.");
        if (quantity > stockQty)
            throw new IllegalStateException(
                    "Insufficient stock for '" + name + "'. Available: " + stockQty + ", requested: " + quantity + ".");
        stockQty -= quantity;
    }

    /** Increase available stock by {@code quantity}. Throws if the quantity is not positive. */
    public void restock(int quantity) {
        if (quantity <= 0)
            throw new IllegalArgumentException("Quantity must be greater than zero.");
        stockQty += quantity;
    }

    /** Sets the initial stock level. Intended for entity creation / seeding only. */
    public void setInitialStock(int quantity) {
        if (quantity < 0)
            throw new IllegalArgumentException("Initial stock cannot be negative.");
        stockQty = quantity;
    }
}
